File: backend/internal/services/drug_service.go
package services

import (
	"encoding/json"
	"errors"
	"strings"
	"time"

	"mediguide/internal/models"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

var ErrDrugInvalidPayload = errors.New("invalid drug payload")

type DrugService struct{ DB *gorm.DB }

type DrugListInput struct {
	Page                  PageInput
	Search                string
	Status                string
	ReviewStatus          string
	DrugClassID           string
	TherapeuticCategoryID string
	Route                 string
	PregnancyCategory     string
	WHOEML                *bool
	Antimicrobial         *bool
	Sort                  string
	Order                 string
}

type DrugInput struct {
	DrugClassID           *string         `json:"drug_class_id"`
	TherapeuticCategoryID *string         `json:"therapeutic_category_id"`
	Name                  *string         `json:"name"`
	BrandNames            *string         `json:"brand_names"`
	Description           *string         `json:"description"`
	MechanismOfAction     *string         `json:"mechanism_of_action"`
	AdultDose             *string         `json:"adult_dose"`
	PediatricDose         *string         `json:"pediatric_dose"`
	ElderlyDose           *string         `json:"elderly_dose"`
	MaxDailyDose          *string         `json:"max_daily_dose"`
	RouteOfAdministration *string         `json:"route_of_administration"`
	Frequency             *string         `json:"frequency"`
	Duration              *string         `json:"duration"`
	Indications           *string         `json:"indications"`
	Contraindications     *string         `json:"contraindications"`
	SideEffects           *string         `json:"side_effects"`
	Warnings              *string         `json:"warnings"`
	MonitoringParameters  *string         `json:"monitoring_parameters"`
	PregnancyCategory     *string         `json:"pregnancy_category"`
	ClinicalNotes         *string         `json:"clinical_notes"`
	Categories            json.RawMessage `json:"categories" swaggertype:"array,string"`
	Tags                  json.RawMessage `json:"tags" swaggertype:"array,string"`
	WHOEMLStatus          *bool           `json:"who_eml_status"`
	AntimicrobialStatus   *bool           `json:"antimicrobial_status"`
	ControlledSubstance   *string         `json:"controlled_substance"`
	Status                *string         `json:"status"`
	ReviewStatus          *string         `json:"review_status"`
	SearchKeywords        *string         `json:"search_keywords"`
	ReferenceText         *string         `json:"reference_text"`
}

func (s DrugService) List(in DrugListInput) (*PageResult[models.Drug], error) {
	page := in.Page.Normalize(20, 100)
	query := s.drugQuery()
	if value := strings.TrimSpace(in.Search); value != "" {
		like := "%" + value + "%"
		query = query.Where(`LOWER(d.name) LIKE LOWER(?) OR LOWER(COALESCE(d.brand_names, '')) LIKE LOWER(?) OR LOWER(COALESCE(d.description, '')) LIKE LOWER(?) OR LOWER(COALESCE(d.indications, '')) LIKE LOWER(?) OR LOWER(COALESCE(d.search_keywords, '')) LIKE LOWER(?)`, like, like, like, like, like)
	}
	for _, filter := range []struct {
		value  string
		column string
	}{
		{in.Status, "d.status"},
		{in.ReviewStatus, "d.review_status"},
		{in.DrugClassID, "d.drug_class_id"},
		{in.TherapeuticCategoryID, "d.therapeutic_category_id"},
		{in.Route, "d.route_of_administration"},
		{in.PregnancyCategory, "d.pregnancy_category"},
	} {
		value, column := filter.value, filter.column
		if strings.TrimSpace(value) != "" {
			query = query.Where(column+" = ?", strings.TrimSpace(value))
		}
	}
	if in.WHOEML != nil {
		query = query.Where("d.who_eml_status = ?", *in.WHOEML)
	}
	if in.Antimicrobial != nil {
		query = query.Where("d.antimicrobial_status = ?", *in.Antimicrobial)
	}
	var total int64
	if err := query.Session(&gorm.Session{}).Distinct("d.id").Count(&total).Error; err != nil {
		return nil, err
	}
	items := []models.Drug{}
	if err := query.Session(&gorm.Session{}).Select("d.*, dc.name AS drug_class_name, tc.name AS therapeutic_category_name").
		Order(drugSort(in.Sort, in.Order)).Limit(page.PerPage).Offset(page.Offset()).Find(&items).Error; err != nil {
		return nil, err
	}
	return NewPageResult(items, page, total), nil
}

func (s DrugService) Get(id uuid.UUID) (*models.Drug, error) {
	var item models.Drug
	err := s.drugQuery().Select("d.*, dc.name AS drug_class_name, tc.name AS therapeutic_category_name").
		Where("d.id = ?", id).Take(&item).Error
	return &item, err
}

func (s DrugService) Create(in DrugInput) (*models.Drug, error) {
	if in.Name == nil || strings.TrimSpace(*in.Name) == "" {
		return nil, ErrDrugInvalidPayload
	}
	drug := models.Drug{Name: strings.TrimSpace(*in.Name), Status: "active", ReviewStatus: "pending"}
	if err := applyDrugInput(&drug, in); err != nil {
		return nil, err
	}
	if err := s.DB.Create(&drug).Error; err != nil {
		return nil, err
	}
	return s.Get(drug.ID)
}

func (s DrugService) Update(id uuid.UUID, in DrugInput) (*models.Drug, error) {
	var drug models.Drug
	if err := s.DB.First(&drug, "id = ?", id).Error; err != nil {
		return nil, err
	}
	if err := applyDrugInput(&drug, in); err != nil {
		return nil, err
	}
	drug.UpdatedAt = time.Now().UTC()
	if err := s.DB.Save(&drug).Error; err != nil {
		return nil, err
	}
	return s.Get(id)
}

func (s DrugService) Delete(id uuid.UUID) error {
	result := s.DB.Delete(&models.Drug{}, "id = ?", id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return gorm.ErrRecordNotFound
	}
	return nil
}

func (s DrugService) RecordUsage(userID, drugID uuid.UUID) (*models.DrugUsageLog, error) {
	if _, err := s.Get(drugID); err != nil {
		return nil, err
	}
	log := models.DrugUsageLog{UserID: userID, DrugID: drugID}
	if err := s.DB.Transaction(func(tx *gorm.DB) error {
		if err := tx.Create(&log).Error; err != nil {
			return err
		}
		return tx.Model(&models.Drug{}).Where("id = ?", drugID).
			UpdateColumn("usage_count", gorm.Expr("usage_count + 1")).Error
	}); err != nil {
		return nil, err
	}
	return &log, nil
}

func (s DrugService) drugQuery() *gorm.DB {
	return s.DB.Table("drugs d").
		Joins("LEFT JOIN drug_classes dc ON dc.id = d.drug_class_id").
		Joins("LEFT JOIN therapeutic_categories tc ON tc.id = d.therapeutic_category_id").
		Where("d.deleted_at IS NULL")
}

func applyDrugInput(drug *models.Drug, in DrugInput) error {
	if in.Name != nil {
		if strings.TrimSpace(*in.Name) == "" {
			return ErrDrugInvalidPayload
		}
		drug.Name = strings.TrimSpace(*in.Name)
	}
	if err := setOptionalUUID(&drug.DrugClassID, in.DrugClassID); err != nil {
		return err
	}
	if err := setOptionalUUID(&drug.TherapeuticCategoryID, in.TherapeuticCategoryID); err != nil {
		return err
	}
	copyDrugString := func(target **string, value *string) {
		if value != nil {
			trimmed := strings.TrimSpace(*value)
			if trimmed == "" {
				*target = nil
			} else {
				*target = &trimmed
			}
		}
	}
	copyDrugString(&drug.BrandNames, in.BrandNames)
	copyDrugString(&drug.Description, in.Description)
	copyDrugString(&drug.MechanismOfAction, in.MechanismOfAction)
	copyDrugString(&drug.AdultDose, in.AdultDose)
	copyDrugString(&drug.PediatricDose, in.PediatricDose)
	copyDrugString(&drug.ElderlyDose, in.ElderlyDose)
	copyDrugString(&drug.MaxDailyDose, in.MaxDailyDose)
	copyDrugString(&drug.RouteOfAdministration, in.RouteOfAdministration)
	copyDrugString(&drug.Frequency, in.Frequency)
	copyDrugString(&drug.Duration, in.Duration)
	copyDrugString(&drug.Indications, in.Indications)
	copyDrugString(&drug.Contraindications, in.Contraindications)
	copyDrugString(&drug.SideEffects, in.SideEffects)
	copyDrugString(&drug.Warnings, in.Warnings)
	copyDrugString(&drug.MonitoringParameters, in.MonitoringParameters)
	copyDrugString(&drug.PregnancyCategory, in.PregnancyCategory)
	copyDrugString(&drug.ClinicalNotes, in.ClinicalNotes)
	copyDrugString(&drug.ControlledSubstance, in.ControlledSubstance)
	copyDrugString(&drug.SearchKeywords, in.SearchKeywords)
	copyDrugString(&drug.ReferenceText, in.ReferenceText)
	if in.Categories != nil {
		if !json.Valid(in.Categories) {
			return ErrDrugInvalidPayload
		}
		drug.CategoriesJSON = datatypes.JSON(in.Categories)
	}
	if in.Tags != nil {
		if !json.Valid(in.Tags) {
			return ErrDrugInvalidPayload
		}
		drug.TagsJSON = datatypes.JSON(in.Tags)
	}
	if in.WHOEMLStatus != nil {
		drug.WHOEMLStatus = *in.WHOEMLStatus
	}
	if in.AntimicrobialStatus != nil {
		drug.AntimicrobialStatus = *in.AntimicrobialStatus
	}
	if in.Status != nil {
		if !oneOf(*in.Status, "active", "inactive", "under_review", "archived") {
			return ErrDrugInvalidPayload
		}
		drug.Status = strings.TrimSpace(*in.Status)
	}
	if in.ReviewStatus != nil {
		if !oneOf(*in.ReviewStatus, "approved", "pending", "needs_update") {
			return ErrDrugInvalidPayload
		}
		drug.ReviewStatus = strings.TrimSpace(*in.ReviewStatus)
	}
	return nil
}

func setOptionalUUID(target **uuid.UUID, raw *string) error {
	if raw == nil {
		return nil
	}
	if strings.TrimSpace(*raw) == "" {
		*target = nil
		return nil
	}
	value, err := uuid.Parse(strings.TrimSpace(*raw))
	if err != nil {
		return ErrDrugInvalidPayload
	}
	*target = &value
	return nil
}

func oneOf(value string, allowed ...string) bool {
	value = strings.TrimSpace(value)
	for _, candidate := range allowed {
		if value == candidate {
			return true
		}
	}
	return false
}

func drugSort(field, order string) string {
	columns := map[string]string{"name": "d.name", "created_at": "d.created_at", "updated_at": "d.updated_at", "usage_count": "d.usage_count"}
	column := columns[strings.TrimSpace(field)]
	if column == "" {
		column = "d.name"
	}
	direction := "ASC"
	if strings.EqualFold(strings.TrimSpace(order), "desc") {
		direction = "DESC"
	}
	return column + " " + direction
}

File: dashboard/services/drug.service.ts
import { backendClient } from "@/lib/backend-client";
import type {
  DrugCategoriesResponse,
  DrugClassesResponse,
  DrugsResponse,
  DrugTagsResponse,
  TherapeuticCategoriesResponse,
} from "@/types/backend-types";
import type {
  ModelsDrug,
  ModelsDrugCategory,
  ModelsDrugClass,
  ModelsDrugTag,
  ModelsTherapeuticCategory,
  ServicesDrugCategoryInput,
  ServicesDrugInput,
  ServicesDrugNamedReferenceInput,
  ServicesDrugTagInput,
} from "@/types/generated/backend-openapi";
import type { DomainPageQuery, DomainPageResult } from "@/types/data-table";

interface Page<T> {
  items: T[];
  page: number;
  per_page: number;
  total_items: number;
  total_pages: number;
}
type JsonInput = Record<string, unknown>;
export type DrugRecord = DrugsResponse<{
  categories: DrugCategoriesResponse[];
  tags: DrugTagsResponse[];
  drug_class: DrugClassesResponse;
  therapeutic_category: TherapeuticCategoriesResponse;
}>;
export type DrugCategoryRecord = DrugCategoriesResponse<{
  parent_category: DrugCategoriesResponse;
}>;

const base = (
  value: { id?: string; created_at?: string; updated_at?: string },
  collection: string,
) => ({
  id: value.id || "",
  created: value.created_at || "",
  updated: value.updated_at || "",
  collectionId: collection,
  collectionName: collection,
});
const normalizeCategory = (v: ModelsDrugCategory) =>
  ({
    ...v,
    ...base(v, "drug_categories"),
    parent_category: v.parent_category_id || "",
    expand: {},
  }) as unknown as DrugCategoryRecord;
const normalizeTag = (v: ModelsDrugTag) =>
  ({ ...v, ...base(v, "drug_tags") }) as DrugTagsResponse;
const normalizeClass = (v: ModelsDrugClass) =>
  ({ ...v, ...base(v, "drug_classes") }) as DrugClassesResponse;
const normalizeTherapeutic = (v: ModelsTherapeuticCategory) =>
  ({
    ...v,
    ...base(v, "therapeutic_categories"),
  }) as TherapeuticCategoriesResponse;
const normalizeDrug = (v: ModelsDrug): DrugRecord =>
  ({
    ...v,
    ...base(v, "drugs"),
    categories: v.categories_json || [],
    tags: v.tags_json || [],
    drug_class: v.drug_class_id || "",
    therapeutic_category: v.therapeutic_category_id || "",
    references: v.reference_text || "",
    usageCount: v.usage_count || 0,
    expand: {
      categories: [],
      tags: [],
      drug_class: { id: v.drug_class_id || "", name: v.drug_class_name || "" },
      therapeutic_category: {
        id: v.therapeutic_category_id || "",
        name: v.therapeutic_category_name || "",
      },
    },
  }) as unknown as DrugRecord;

const drugPayload = (v: JsonInput): ServicesDrugInput => {
  const allowed = [
    "name",
    "description",
    "brand_names",
    "adult_dose",
    "pediatric_dose",
    "elderly_dose",
    "frequency",
    "duration",
    "max_daily_dose",
    "route_of_administration",
    "pregnancy_category",
    "controlled_substance",
    "who_eml_status",
    "antimicrobial_status",
    "mechanism_of_action",
    "indications",
    "contraindications",
    "side_effects",
    "warnings",
    "monitoring_parameters",
    "clinical_notes",
    "search_keywords",
    "status",
    "review_status",
  ] as const;
  const payload: Record<string, unknown> = {};
  for (const key of allowed) if (v[key] !== undefined) payload[key] = v[key];
  payload.categories =
    (v.categories as string[] | undefined) ||
    (v.categories_json as string[] | undefined);
  payload.tags =
    (v.tags as string[] | undefined) || (v.tags_json as string[] | undefined);
  payload.drug_class_id =
    (v.drug_class_id as string | undefined) ||
    (v.drug_class as string | undefined);
  payload.therapeutic_category_id =
    (v.therapeutic_category_id as string | undefined) ||
    (v.therapeutic_category as string | undefined);
  payload.reference_text =
    (v.reference_text as string | undefined) ||
    (v.references as string | undefined);
  return payload as ServicesDrugInput;
};

const categoryPayload = (data: JsonInput): ServicesDrugCategoryInput => ({
  name: data.name as string | undefined,
  description: data.description as string | undefined,
  color: data.color as string | undefined,
  icon: data.icon as string | undefined,
  sort_order: data.sort_order as number | undefined,
  status: data.status as string | undefined,
  parent_category_id:
    (data.parent_category_id as string | undefined) ||
    (Array.isArray(data.parent_category)
      ? (data.parent_category[0] as string | undefined)
      : (data.parent_category as string | undefined)),
});

async function list<T>(path: string, query: Record<string, unknown> = {}) {
  return backendClient.send<Page<T>>(path, {
    query: { page: 1, per_page: 100, ...query },
  });
}
const table =
  <TWire, TRecord>(
    fetcher: (q: Record<string, unknown>) => Promise<Page<TWire>>,
    normalize: (v: TWire) => TRecord,
  ) =>
  async (q: DomainPageQuery): Promise<DomainPageResult<TRecord>> => {
    const r = await fetcher({
      page: q.page,
      per_page: q.perPage,
      search: q.search || undefined,
    });
    return {
      items: r.items.map(normalize),
      page: r.page,
      perPage: r.per_page,
      totalItems: r.total_items,
      totalPages: r.total_pages,
    };
  };

async function hydrateDrugRelations(
  records: DrugRecord[],
): Promise<DrugRecord[]> {
  const [categories, tags] = await Promise.all([
    list<ModelsDrugCategory>("/api/v2/drug-categories", { per_page: 100 }),
    list<ModelsDrugTag>("/api/v2/drug-tags", { per_page: 100 }),
  ]);
  const categoryIndex = new Map(
    categories.items.map((v) => [v.id, normalizeCategory(v)]),
  );
  const tagIndex = new Map(tags.items.map((v) => [v.id, normalizeTag(v)]));
  return records.map(
    (record) =>
      ({
        ...record,
        expand: {
          ...record.expand,
          categories: (record.categories || [])
            .map((id) => categoryIndex.get(id))
            .filter(Boolean),
          tags: (record.tags || [])
            .map((id) => tagIndex.get(id))
            .filter(Boolean),
        },
      }) as DrugRecord,
  );
}

export const drugService = {
  async list(query: Record<string, unknown> = {}) {
    return list<ModelsDrug>("/api/v2/drugs", query);
  },
  async listTable(query: DomainPageQuery) {
    const r = await this.list({
      page: query.page,
      per_page: query.perPage,
      search: query.search || undefined,
      sort: "name",
      order: "asc",
    });
    return {
      items: await hydrateDrugRelations(r.items.map(normalizeDrug)),
      page: r.page,
      perPage: r.per_page,
      totalItems: r.total_items,
      totalPages: r.total_pages,
    };
  },
  async get(id: string) {
    return (
      await hydrateDrugRelations([
        normalizeDrug(
          await backendClient.send<ModelsDrug>(`/api/v2/drugs/${id}`),
        ),
      ])
    )[0];
  },
  async create(data: JsonInput) {
    return normalizeDrug(
      await backendClient.send<ModelsDrug>("/api/v2/drugs", {
        method: "POST",
        body: JSON.stringify(drugPayload(data)),
      }),
    );
  },
  async update(id: string, data: JsonInput) {
    return normalizeDrug(
      await backendClient.send<ModelsDrug>(`/api/v2/drugs/${id}`, {
        method: "PATCH",
        body: JSON.stringify(drugPayload(data)),
      }),
    );
  },
  async delete(id: string) {
    await backendClient.send<void>(`/api/v2/drugs/${id}`, { method: "DELETE" });
  },
};

export const drugReferenceService = {
  async listCategories(query: Record<string, unknown> = {}) {
    return list<ModelsDrugCategory>("/api/v2/drug-categories", query);
  },
  listCategoriesTable: table(
    (q) => list<ModelsDrugCategory>("/api/v2/drug-categories", q),
    normalizeCategory,
  ),
  async allCategories() {
    return (
      await list<ModelsDrugCategory>("/api/v2/drug-categories", {
        per_page: 100,
        status: "active",
      })
    ).items.map(normalizeCategory);
  },
  async createCategory(data: JsonInput) {
    return normalizeCategory(
      await backendClient.send<ModelsDrugCategory>("/api/v2/drug-categories", {
        method: "POST",
        body: JSON.stringify(categoryPayload(data)),
      }),
    );
  },
  async updateCategory(id: string, data: JsonInput) {
    return normalizeCategory(
      await backendClient.send<ModelsDrugCategory>(
        `/api/v2/drug-categories/${id}`,
        { method: "PATCH", body: JSON.stringify(categoryPayload(data)) },
      ),
    );
  },
  async deleteCategory(id: string) {
    await backendClient.send<void>(`/api/v2/drug-categories/${id}`, {
      method: "DELETE",
    });
  },
  async listTags(query: Record<string, unknown> = {}) {
    return list<ModelsDrugTag>("/api/v2/drug-tags", query);
  },
  listTagsTable: table(
    (q) => list<ModelsDrugTag>("/api/v2/drug-tags", q),
    normalizeTag,
  ),
  async allTags() {
    return (
      await list<ModelsDrugTag>("/api/v2/drug-tags", {
        per_page: 100,
        status: "active",
      })
    ).items.map(normalizeTag);
  },
  async createTag(data: ServicesDrugTagInput) {
    return normalizeTag(
      await backendClient.send<ModelsDrugTag>("/api/v2/drug-tags", {
        method: "POST",
        body: JSON.stringify(data),
      }),
    );
  },
  async updateTag(id: string, data: ServicesDrugTagInput) {
    return normalizeTag(
      await backendClient.send<ModelsDrugTag>(`/api/v2/drug-tags/${id}`, {
        method: "PATCH",
        body: JSON.stringify(data),
      }),
    );
  },
  async deleteTag(id: string) {
    await backendClient.send<void>(`/api/v2/drug-tags/${id}`, {
      method: "DELETE",
    });
  },
  async allClasses() {
    return (
      await list<ModelsDrugClass>("/api/v2/drug-classes", {
        per_page: 100,
        status: "active",
      })
    ).items.map(normalizeClass);
  },
  async createClass(data: ServicesDrugNamedReferenceInput) {
    return normalizeClass(
      await backendClient.send<ModelsDrugClass>("/api/v2/drug-classes", {
        method: "POST",
        body: JSON.stringify(data),
      }),
    );
  },
  async allTherapeuticCategories() {
    return (
      await list<ModelsTherapeuticCategory>("/api/v2/therapeutic-categories", {
        per_page: 100,
        status: "active",
      })
    ).items.map(normalizeTherapeutic);
  },
  async createTherapeuticCategory(data: ServicesDrugNamedReferenceInput) {
    return normalizeTherapeutic(
      await backendClient.send<ModelsTherapeuticCategory>(
        "/api/v2/therapeutic-categories",
        { method: "POST", body: JSON.stringify(data) },
      ),
    );
  },
};

export const drugCategoryCrud = {
  create: drugReferenceService.createCategory.bind(drugReferenceService),
  update: drugReferenceService.updateCategory.bind(drugReferenceService),
  delete: drugReferenceService.deleteCategory.bind(drugReferenceService),
};
export const drugTagCrud = {
  create: drugReferenceService.createTag.bind(drugReferenceService),
  update: drugReferenceService.updateTag.bind(drugReferenceService),
  delete: drugReferenceService.deleteTag.bind(drugReferenceService),
};
